Move LocationType to SQL

esoergel opened the pull request:
@twymer
(although @czue would probably be interested as well)
Move location types to SQL.  I'm gonna test this out on staging and have Christy QA it before merge.

esoergel committed:
Convert location_types to SQL

esoergel committed:
Remove couch location types

esoergel committed:
Update references to old-style LocationType

esoergel committed:
Update custom reports with new location types

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -100,37 +99,59 @@ def page_context(self):
+        def mk_loctype(name, code, allowed_parents, administrative):
+            parents = allowed_parents
+            if len(parents) > 1:
+                messages.warning(request, _("Location types cannot have "
```
F823 local variable '_' (defined in enclosing scope on line 11) referenced before assignment

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -100,37 +99,59 @@ def page_context(self):
+            location_type.parent_type = parent
+            location_type.save()
+
+        #TODO add server-side input validation here (currently validated on client)
```
E265 block comment should start with '# '

dimagimon commented on custom/ilsgateway/api.py:
```diff
@@ -126,16 +125,15 @@ class ILSGatewayAPI(APISynchronization):
+                domain=self.domain,
+                name=loc_type,
+                parent_type=previous,
+                administrative=(value != 'FACILITY'),
```
F821 undefined name 'value'

esoergel replied:
you win this round, @dimagimon....

twymer commented on corehq/apps/commtrack/util.py:
```diff
@@ -149,36 +148,22 @@ def _needs_commtrack_config(domain):
+            administrative=administrative,
+        )
+        location_type.save()
+        previous = location_type
```
Ahhh.. I might actually be removing this after recent feedback from Sheel that we shouldn't populate bogus types.

twymer commented on corehq/apps/domain/models.py:
```diff
@@ -334,6 +332,9 @@ def wrap(cls, data):
 
+        if 'location_types' in data:
+            del data['location_types']
+
```
Shouldn't this get manually cleaned up in the migration? If not, should probably put a comment on what is essentially a handler for legacy data.

twymer commented on corehq/apps/domain/models.py:
```diff
@@ -1024,11 +1025,17 @@ def published_by(self):
+    @property
+    def location_types(self):
+        from corehq.apps.locations.models import LocationType
+        return LocationType.objects.filter(domain=self.name).all()
```
:heart:

twymer commented on corehq/apps/locations/models.py:
```diff
@@ -409,16 +436,7 @@ def linked_supply_point(self):
-        for loc_type in Domain.get_by_name(self.domain).location_types:
-            if loc_type.name == self.location_type:
-                return loc_type
+        return self.sql_location.location_type
```
:heart_eyes:

czue replied:
:+1:

twymer commented:
me: "should be easy to just port it to sql"

sorry!

esoergel commented:
Test failure is real, I thought I'd gone in and added a teardown method, but I must've forgotten.

esoergel committed:
PEP8 cleanup

esoergel committed:
Fix migration (sorta)

czue commented:
what's up with this PR name?

esoergel commented:
The impetus for moving location types to SQL is to store stock rate information on each location type.  I wanted to PR the SQL conversion separately though, so that name isn't the most accurate.  I'll change it.

esoergel committed:
Make location tests pass

dimagimon commented:
:+1: No lint errors found.

esoergel committed:
Remove location_types check

dimagimon commented:
:+1: No lint errors found.

esoergel committed:
Make migration not dependant upon external code

dimagimon commented:
:+1: No lint errors found.

esoergel committed:
Add location_type to Django admin

esoergel committed:
Add location_types sorter

czue commented on corehq/apps/commtrack/management/commands/check_multiple_parentage.py:
```diff
@@ -1,44 +0,0 @@
-from django.core.management.base import BaseCommand
```
not a big deal, but in a PR like this if you're just deleting cruft I always prefer to just submit that separately

czue commented:
@esoergel can you merge this with master? I just gave it a quick shot and there's a lot of conflicts

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into loc-type-stock-rates

Conflicts:
	corehq/apps/commtrack/util.py
	corehq/apps/domain/models.py
	corehq/apps/locations/models.py
	corehq/apps/locations/schema.py
	corehq/apps/locations/tests/test_locations.py
	custom/ewsghana/api.py

esoergel committed:
Fix merge conflicts from locations-as-groups

esoergel committed:
Move shares_cases and view_descendents to SQL

dimagimon commented:
There are 6 errors:

* corehq/apps/commtrack/tests/util.py, line 9 - F401 'LocationType' imported but unused
* corehq/apps/commtrack/tests/util.py, line 13 - E126 continuation line over-indented for hanging indent
* corehq/apps/locations/models.py, line 42 - E713 test for membership should be 'not in'
* corehq/apps/locations/models.py, line 56 - E301 expected 1 blank line, found 0
* corehq/apps/locations/tests/test_location_types.py, line 9 - E301 expected 1 blank line, found 0
* corehq/apps/locations/tests/test_locations.py, line 1 - E501 line too long (128 > 115 characters)

esoergel committed:
PEP8

dimagimon commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
+        Sorts location types by heirarchy
+        """
+        ordered_loc_types = []
+        def step_through_graph(heirarchy):
```
E301 expected 1 blank line, found 0

dimagimon replied:
E301 expected 1 blank line, found 0

esoergel committed:
Merge remote-tracking branch 'origin/master' into loc-type-stock-rates

Conflicts:
	corehq/apps/locations/tests/__init__.py

czue commented:
@esoergel got this error trying to migrate locally, not sure if just an issue with my environment though: http://pastebin.com/c9G9gMNU

czue commented on corehq/apps/locations/migrations/0009_populate_location_type_from_string.py:
```diff
@@ -0,0 +1,162 @@
+
+    @memoized
+    def domain_loc_types(self, domain):
+        domain = get_db().view("domain/domains", key=domain, reduce=False,
```
not a big deal but should be `Domain.get_db`

czue commented on corehq/apps/locations/migrations/0009_populate_location_type_from_string.py:
```diff
@@ -0,0 +1,162 @@
+            elif code in loc_types:
+                couch_loc_type = loc_types[code]
+            else:
+                couch_loc_type = loc_types.values()[0]
```
this should probably fail hard

czue commented on corehq/apps/locations/migrations/0009_populate_location_type_from_string.py:
```diff
@@ -0,0 +1,162 @@
+            loc_type = self.orm.LocationType.objects.get(domain=domain, code=code)
+        except self.orm.LocationType.DoesNotExist:
+            loc_types = self.domain_loc_types(domain)
+            if not loc_types:
```
if no location types in the domain this should fail louder

czue commented on corehq/apps/locations/migrations/0009_populate_location_type_from_string.py:
```diff
@@ -0,0 +1,162 @@
+
+            parents = couch_loc_type['allowed_parents']
+            if parents and parents[0]:
+                # I sure hope there are no cycles...
```
me too

czue commented on corehq/apps/locations/migrations/0009_populate_location_type_from_string.py:
```diff
@@ -0,0 +1,162 @@
+                couch_loc_type = loc_types.values()[0]
+
+            parents = couch_loc_type['allowed_parents']
+            if parents and parents[0]:
```
should assert this is exactly one element or log an error

benrudolph commented on corehq/apps/locations/views.py:
```diff
@@ -118,23 +120,43 @@ def _get_loctype_info(self, loctype):
+            if cleaned_code != code:
+                err = _('Location type code "{code}" is invalid. No spaces or '
+                        'special characters are allowed. It has been replaced '
+                        'with "{new_code}".')
```
small, but it seems `code` is unicode and it could throw an error being put into the bytestring

esoergel replied:
I think the name is misleading, looks like unicode_slug accepts unicode, but the output is ascii (it calls out to [this django function](https://github.com/django/django/blob/master/django/utils/text.py#L413-L423))

benrudolph replied:
ah ok got it

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into loc-type-stock-rates

Conflicts:
	corehq/apps/domain/models.py

esoergel committed:
Merge remote-tracking branch 'origin/master' into loc-type-stock-rates

esoergel committed:
make loc_types migration non-destructive

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into loc-type-stock-rates

Conflicts:
	custom/ilsgateway/api.py

esoergel committed:
Allow null code values in initial migration

esoergel committed:
Also add loc_type names to mapping

esoergel committed:
Coerce undefined values to a boolean

dimagimon commented on corehq/apps/locations/util.py:
```diff
@@ -62,13 +62,14 @@ def loc_to_json(loc):
-    return [(loc_type.name, [p or None for p in loc_type.allowed_parents])
-            for loc_type in Domain.get_by_name(domain).location_types]
+def location_type_tree(domain):
+    loc_types = Domain.get_by_name(domain).location_types
```
F841 local variable 'loc_types' is assigned to but never used

esoergel committed:
Use the LocationTypeManager sorting method

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into loc-type-stock-rates

Conflicts:
	custom/ewsghana/api.py

dannyroberts commented:
Talked it over with @esoergel, plan is to merge (right) before monday deploy, and have him on hand when I deploy.

czue commented:
(note that I haven't actually reviewed this yet, but fine if someone else has)

@esoergel can you confirm that the migration is not destructive? as in - if something goes wrong we can just revert and the database will be in a usable state off the old code?

czue commented:
also, confirming QA happened?

czue commented:
(also, fyi, not mergeable anymore)

esoergel commented:
I'll resolve the merge conflict.

The information is not lost, but I don't think I can confidently say that the code will just work if we revert -
 1. If any domains are saved while running this branch, it will rename the domain.location_type property to domain.obsolete_location_type.
 2. The location_types property on SQLLocation is converted from a string to a foreign key (although a proper reverse migration can make that reversable - I'll write that in).  In either case, the information is still there on the couch Location, and a save in the old code would populate that value.

I'll offline, but let's count on not merging this today.

dannyroberts commented:
@esoergel let me know if you offlined what you decided, or else grab me tomorrow to figure this out

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -102,14 +101,18 @@ def page_context(self):
-            'code': loctype.code,
-            'allowed_parents': [p or None for p in loctype.allowed_parents],
+            'parent_pk': (loctype.parent_type.pk
+                            if loctype.parent_type else None),
```
E127 continuation line over-indented for visual indent

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -118,23 +121,34 @@ def _get_loctype_info(self, loctype):
-                    'It has been replaced with "{new_code}".'
+        def mk_loctype(loc_type):
+            print json.dumps(loc_type, indent=4)
+            pk = loc_type['pk']
```
F841 local variable 'pk' is assigned to but never used

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into loc-type-stock-rates

Conflicts:
	corehq/apps/groups/models.py

esoergel committed:
Temporarily remove "Remove" button from loc types

This allows the user to get into an inconsistent state, so I'm disabling
it until we have a proper removal workflow

esoergel committed:
Make Location.location_type migration reversible

esoergel committed:
Actually update code after validation

esoergel committed:
Fix misc location types UI bugs

esoergel committed:
update ilsgateway location type test

dannyroberts commented on corehq/apps/groups/models.py:
```diff
@@ -181,10 +181,7 @@ def get_case_sharing_groups(cls, domain, wrap=True):
-                # fixes http://manage.dimagi.com/default.asp?158564
-                if getattr(location.couch_location().location_type_object,
-                           'shares_cases', None)
+                if location.couch_location.location_type_object.shares_cases
```
Did this just get lost in a merge somewhere, or did you intentionally change it back from my hotfix because it's no longer needed?

esoergel replied:
That was intentional.  There are no longer any locations without location types, and it's not possible to make locations without location types, so this shouldn't fail.  However, this isn't (yet) enforced at the db schema level (although I plan on doing that in a separate PR), so I could be talked in to putting this back in.

snopoke replied:
@esoergel Do you realize that this goes SQLLocation -> CouchLocation -> SQLLocation.

I think `location.location_type.shares_cases` is equivalent but without the extra Couch lookup

snopoke replied:
You could also eagerly fetch the `location_type` with the location to avoid extra SQL queries

esoergel replied:
Haha, wow yeah that's not the way to do it.  This is a relic of when the `SQLLocation.location_type` was just a string, and this was the simplest way to get the location type object.  I'll update it.

esoergel replied:
https://github.com/dimagi/commcare-hq/pull/6574

czue commented on corehq/apps/commtrack/util.py:
```diff
@@ -88,6 +87,27 @@ def get_or_create_default_program(domain):
+            defaults={
+                'parent_type': previous,
+                'administrative': administrative,
+            },
```
neat - i didn't know about `defaults`. https://docs.djangoproject.com/en/1.7/ref/models/querysets/#django.db.models.query.QuerySet.get_or_create

benrudolph replied:
that is cool. looks like that definitive guide to django you read is paying off

czue commented on corehq/apps/commtrack/util.py:
```diff
@@ -88,6 +87,27 @@ def get_or_create_default_program(domain):
+        previous = location_type
+
+
+
```
don't care, but surprised @dimagimon didn't yell about whitespace.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
 
 
+class LocationTypeManager(models.Manager):
+    def full_heirarchy(self, domain):
```
(tiny) hierarchy is misspelled throughout this file

esoergel replied:
Wow.  Updating PR.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
+
+    def by_domain(self, domain):
+        """
+        Sorts location types by heirarchy
```
it's not obvious to me what the order is if there are locations that have multiple child types. i think it's analogous to a depth first search? out of curiosity, what's the use case for this ordering?

esoergel replied:
yeah, it's like depth-first search.  I'm using this ordering on the edit location types page, before I added it, they just appeared in an arbitrary order, which looks confusing when there's a strict ~~heirarchy~~ hierarchy implied.  This is arguably unneeded extra computation if you needed to get all LocationTypes and don't care about order, but I don't think that's something that really happens except when you're going to display them to the user.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
+
+class LocationType(models.Model):
+    domain = models.CharField(max_length=255, db_index=True)
+    name = models.CharField(max_length=255)
```
assume we've checked for null names?

esoergel replied:
That wasn't directly part of the checking, no.  I just checked though, and there aren't any.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
+class LocationType(models.Model):
+    domain = models.CharField(max_length=255, db_index=True)
+    name = models.CharField(max_length=255)
+    code = models.SlugField(db_index=False, null=True)
```
(why is this `null`?)

esoergel replied:
There are a bunch of existing location types where this is null (old stuff, I think), and this column isn't actually used anywhere.  I'm going to get rid of it eventually.

czue commented on corehq/apps/locations/migrations/0009_populate_location_type_from_string.py:
```diff
@@ -0,0 +1,150 @@
+                defaults={
+                    'name': couch_loc_type['name'],
+                    'parent_type': parent_type,
+                    'administrative': couch_loc_type['administrative'] or False,
```
am I missing something or do we need to add `shares_cases` and `view_descendants` here?

esoergel replied:
No, you're right

czue commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
+    def save(self, *args, **kwargs):
+        if not self.code:
+            from corehq.apps.commtrack.util import unicode_slug
+            self.code = unicode_slug(self.name)
```
assume this is some legacy thing? might be good to have a comment. this also seems to imply that `code` should be `null=True`

esoergel replied:
Yeah, this is a legacy thing, it was on the old LocationType model, I assume part of a lazy migration.  In retrospect, it probably would've been better to put this logic in the data migration.  I think I can just remove it when I take out `code`.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
     name = models.CharField(max_length=100, null=True)
     location_id = models.CharField(max_length=100, db_index=True, unique=True)
-    location_type = models.CharField(max_length=255)
+    location_type = models.ForeignKey(LocationType, null=True)
```
surprised by `null=True` here too

esoergel replied:
This one is surprisingly tricky, a whole bunch of the commtrack and locations tests create locations without types or with invalid types.  I have another branch where I make this non-null, hoping to push that through afterwards separately.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -154,18 +219,11 @@ def reporting_group_object(self, user_id=None):
-        If there are no products specified for this location, assume all
-        products for the domain are relevant.
-        """
-        return self.products.all() or SQLProduct.by_domain(self.domain)
```
is this just not called anywhere?

esoergel replied:
That's correct.

czue commented on corehq/apps/locations/models.py:
```diff
@@ -254,6 +311,12 @@ def _sync_location(self):
+            sql_location.location_type, _ = LocationType.objects.get_or_create(
+                domain=self.domain,
+                name=self.location_type,
+            )
```
when do we anticipate that we will be creating a location without a valid type? i don't quite see how this should be possible, and it seems like this could just allow us to introduce other bugs in the code.

esoergel replied:
I had actually corrected this, but that's on a different branch (the one with the non-null constraint) was having some trouble getting it past the tests, which create locations without types left and right.

czue commented on corehq/apps/locations/templates/locations/settings.html:
```diff
@@ -71,7 +71,7 @@
                             <th>{% trans "Shares Cases" %}</th>
                             <th>{% trans "View Child Cases" %}</th>
-                            <th></th>
+                            {# <th></th> #}
```
should the changes in this file be deletions instead of comments?

esoergel replied:
This is intended to be a temporary fix, so I put it in as comments.

czue commented on corehq/apps/locations/tests/test_location_types.py:
```diff
@@ -0,0 +1,54 @@
+        self.domain.delete()
+        LocationType.objects.filter(domain=self.domain.name).delete()
+
+    def test_heirarchy(self):
```
nice to see tests. might be good to add one for `by_domain` as well.

czue commented on corehq/apps/locations/views.py:
```diff
@@ -102,42 +100,121 @@ def page_context(self):
-            return LocationType(**loctype)
-
-        #TODO add server-side input validation here (currently validated on client)
+        def mk_loctype(name, code, allowed_parents, administrative,
```
any reason for this to take in a list still? might be nice to change the function to make it explicit that there's only at most one

czue commented on corehq/apps/locations/views.py:
```diff
@@ -102,42 +100,121 @@ def page_context(self):
+        def mk_loctype(name, code, allowed_parents, administrative,
+                       shares_cases, view_descendants):
+            if allowed_parents and allowed_parents[0]:
+                parent = sql_loc_types[allowed_parents[0]]
```
(this seems like it could be hiding bugs if you passed in more than one allowed parent - another possible motivator for changing the function api)

czue commented on corehq/apps/locations/views.py:
```diff
@@ -102,42 +100,121 @@ def page_context(self):
+    # This is largely copy-pasted from the LocationTypeManager
+    def get_heirarchy(self, loc_types):
+        """
+        Return loc types in order from parents to
```
looks like you forgot to finish your

benrudolph replied:
lol

czue commented on corehq/apps/locations/views.py:
```diff
@@ -102,42 +100,121 @@ def page_context(self):
+                assert lt['name'] not in visited, \
+                    "There's a loc type cycle, we need to prohibit that"
+                visited.add(lt['name'])
+                parents = lt['allowed_parents']
```
seems like this could benefit from an assert that `allowed_parents` is at most one and then you can modify the `mk_loctype` accordingly

czue commented on corehq/apps/locations/views.py:
```diff
@@ -102,42 +100,121 @@ def page_context(self):
+            """
+            name = loc_type['name']
+            parents = loc_type['allowed_parents']
+            parent = lt_dict.get(parents[0], None) if parents else None
```
or here

czue commented on corehq/apps/locations/views.py:
```diff
@@ -102,42 +100,121 @@ def page_context(self):
+            parents = loc_type['allowed_parents']
+            parent = lt_dict.get(parents[0], None) if parents else None
+            if not parent:
+                lt_heirarchy = heirarchy
```
(spelling mistakes in this file too)

czue commented on custom/ewsghana/api.py:
```diff
@@ -178,40 +178,99 @@ def _create_supply_point_from_location(self, supply_point, location):
+            name="facility",
+            administrative=False,
+            parent_type=district,
+        )
```
(not actionable) it might be more readable if you just made a datastructure to hold the whole structure. i realize that would be a little annoying because you'd have to make and map the parents first though.

czue commented:
@esoergel done with my pass. looks really great overall. I think there are a few places where you could clean things up, but i'm open to push back. the main things i'd like to sort out before merging are this comment: https://github.com/dimagi/commcare-hq/pull/5511/files#r26800672 and just clarification on what is expected to be null in the models.

esoergel committed:
heirarchy -> hierarchy

dimagimon commented on corehq/apps/locations/models.py:
```diff
@@ -9,20 +9,85 @@
+        Sorts location types by hierarchy
+        """
+        ordered_loc_types = []
+        def step_through_graph(hierarchy):
```
E301 expected 1 blank line, found 0

esoergel commented:
Thanks for the review, Cory.  I'd like to chat about a couple of these in person, I'll come find you in the office today.

esoergel committed:
Also copy over new LocationType properties

dannyroberts commented:
(not mergeable)

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into loc-type-stock-rates

Conflicts:
	custom/ewsghana/api.py

esoergel committed:
Clean up ewsghana location types

esoergel committed:
Allow users to delete location-less location types

esoergel commented:
:tada: :balloon: :cake: :birthday: :fireworks:

benrudolph commented:
been watching this baby grow! you are a :star2:

twymer commented:
Glad this rabbit hole is finally closed :heart: :blue_heart: :green_heart: :yellow_heart: :purple_heart: :sparkling_heart: :revolving_hearts: :two_hearts:

czue commented:
@TylerSheffels @twymer 

┓┏┓┏┓┃
┛┗┛┗┛┃＼○／
┓┏┓┏┓┃  /      Friday
┛┗┛┗┛┃ノ)
┓┏┓┏┓┃         deploy,
┛┗┛┗┛┃
┓┏┓┏┓┃         good
┛┗┛┗┛┃
┓┏┓┏┓┃         luck!
┃┃┃┃┃┃
┻┻┻┻┻┻
